# Laboratorium 3 — NumPy
**Autor:** Katia Kolanko

Ćwiczenie 1: Tworzenie tablic

Import

In [2]:
import numpy as np

Tworzenie tablic

In [2]:
# 1. Z listy
oceny = np.array([4.0, 3.5, 5.0, 4.5, 3.0, 4.0, 5.0])
print(f"Oceny: {oceny}")
print(f"Typ: {oceny.dtype}")

Oceny: [4.  3.5 5.  4.5 3.  4.  5. ]
Typ: float64


In [3]:
# 2. Zera — 10 elementów
zera = np.zeros(10)
print(f"Zera: {zera}")

Zera: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [4]:
# 3. Jedynki — macierz 3×5
jedynki = np.ones((3, 5))
print(f"Jedynki:\n{jedynki}")

Jedynki:
[[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]]


In [5]:
# 4. Sekwencja od 0 do 100, co 10
dziesiatki = np.arange(0, 101, 10)
print(f"Dziesiątki: {dziesiatki}")

Dziesiątki: [  0  10  20  30  40  50  60  70  80  90 100]


In [6]:
# 5. 7 równomiernych punktów od 0 do 1
ulamki = np.linspace(0, 1, 7)
print(f"Ulamki: {ulamki}")

Ulamki: [0.         0.16666667 0.33333333 0.5        0.66666667 0.83333333
 1.        ]


In [7]:
# 6. 20 losowych ocen od 2 do 5
np.random.seed(42)
losowe_oceny = np.random.randint(2, 6, size=20)
print(f"Losowe oceny: {losowe_oceny}")

Losowe oceny: [4 5 2 4 4 5 2 2 4 3 4 4 4 4 5 2 5 5 5 4]


Atrybuty

In [8]:
# Zbadaj tablicę losowych ocen
print(f"shape: {losowe_oceny.shape}")
print(f"ndim: {losowe_oceny.ndim}")
print(f"size: {losowe_oceny.size}")
print(f"dtype: {losowe_oceny.dtype}")

shape: (20,)
ndim: 1
size: 20
dtype: int32


Ćwiczenie 2: Indeksowanie i slicing

Część A — Tablica 1D

In [9]:
miesiace_sprzedaz = np.array([120, 135, 98, 142, 167, 189, 201, 178, 156, 143, 198, 221])

1. Sprzedaż w styczniu (indeks 0):

In [16]:
# Podpowiedź: tablica[indeks]
styczen = miesiace_sprzedaz[0]
print(styczen)

120


2. Sprzedaż w grudniu (ostatni element):

In [15]:
# Podpowiedź: tablica[-1]
grudzien = miesiace_sprzedaz[-1]
print(grudzien)

221


3. Sprzedaż w Q1 (styczeń-marzec, indeksy 0-2):

In [13]:
# Podpowiedź: tablica[start:stop] — stop NIE jest włączony
q1 = miesiace_sprzedaz[0:3]
print(q1)

[120 135  98]


4. Sprzedaż w Q4 (październik-grudzień, ostatnie 3):

In [17]:
# Podpowiedź: tablica[-3:]
q4 = miesiace_sprzedaz[-3:]
print(q4)

[143 198 221]


5. Sprzedaż co drugi miesiąc (styczeń, marzec, maj…):

In [18]:
# Podpowiedź: tablica[::krok]
co_drugi = miesiace_sprzedaz[::2]
print(co_drugi)

[120  98 167 201 156 198]


6. Sprzedaż od tyłu (grudzień → styczeń):

In [19]:
# Podpowiedź: tablica[::-1]
od_tylu = miesiace_sprzedaz[::-1]
print(od_tylu)

[221 198 143 156 178 201 189 167 142  98 135 120]


Część B — Macierz 2D

Sprzedaż 4 produktów w 3 kwartałach

In [3]:
dane = np.array([
[120, 145, 167], # Produkt A
[89, 102, 95], # Produkt B
[234, 267, 289], # Produkt C
[56, 78, 91], # Produkt D
])
produkty = ['Produkt A', 'Produkt B', 'Produkt C', 'Produkt D']
kwartaly = ['Q1', 'Q2', 'Q3']

1. Sprzedaż Produktu A w Q2 (wiersz 0, kolumna 1):

In [22]:
# Podpowiedź: macierz[wiersz, kolumna]
q2 = dane[0,1]
print(q2)

145


2. Cała sprzedaż Produktu C (wiersz 2, wszystkie kolumny):

In [23]:
# Podpowiedź: macierz[wiersz] lub macierz[wiersz, :]
produkt_c = dane[2]
print(produkt_c)

[234 267 289]


3. Sprzedaż wszystkich produktów w Q1 (wszystkie wiersze, kolumna 0):

In [25]:
# Podpowiedź: macierz[:, kolumna]
q1 = dane[:, 0]
print(q1)

[120  89 234  56]


4. Sprzedaż Produktów B i C w Q2 i Q3 (wiersze 1-2, kolumny 1-2):

In [26]:
# Podpowiedź: macierz[w1:w2, k1:k2] — stop NIE jest włączony
bc_q2q3 = dane[1:3, 1:3]
print(bc_q2q3)

[[102  95]
 [267 289]]


Ćwiczenie 3: Operacje wektorowe — analiza biznesowa

In [4]:
# Produkty w sklepie internetowym
nazwy = ['Laptop', 'Tablet', 'Smartfon', 'Słuchawki', 'Ładowarka',
'Klawiatura', 'Mysz', 'Monitor', 'Kamera', 'Drukarka']
ceny = np.array([3499, 1899, 2499, 349, 89, 249, 129, 1499, 599, 899])
ilosci_magazyn = np.array([15, 42, 78, 156, 312, 89, 201, 23, 45, 34])
oceny_klientow = np.array([4.7, 4.2, 4.5, 4.8, 3.9, 4.1, 4.3, 4.6, 3.8, 3.5])

Zadanie 1: Podstawowe statystyki cenowe

In [6]:
# Średnia cena, najdroższy produkt, najtańszy produkt
# Podpowiedzi: .mean(), .max(), .min(), .argmax(), .argmin()

srednia_cena = ceny.mean()
najdrozszy_produkt = ceny.max()
najtanszy_produkt = ceny.min()

print(srednia_cena)
print(najdrozszy_produkt)
print(najtanszy_produkt)

1171.0
3499
89


Średnia cena produktu wynosi 1171 zł, najdroższy produkt kosztuje 3499 zł natomiast najtańszy 89 zł

Zadanie 2: Wartość magazynu

In [8]:
# Oblicz wartość każdego produktu (cena × ilość) i łączną wartość magazynu
# Podpowiedź: mnożenie tablic element-wise

wszystkie_produkty = ceny * ilosci_magazyn
magazyn = wszystkie_produkty.sum()

print(wszystkie_produkty)
print(magazyn)

[ 52485  79758 194922  54444  27768  22161  25929  34477  26955  30566]
549465


Łączna wartość magazynu wynosi 549465zł.

Zadanie 3: Analiza cenowa

In [10]:
# Stworz array z cenami po rabacie 15% (promocja świąteczna)
# Stworz array z cenami po uwzględnieniu VAT 23%
# Ile produktów kosztuje powyżej 1000 zł?
# Podpowiedź: rabat = ceny * 0.85, VAT = ceny * 1.23, zliczanie: (warunek).sum()

rabat = ceny * 0.85
vat = ceny * 1.23

powyzej1000 = (ceny > 1000).sum()

print(rabat)
print(vat)
print(powyzej1000)

[2974.15 1614.15 2124.15  296.65   75.65  211.65  109.65 1274.15  509.15
  764.15]
[4303.77 2335.77 3073.77  429.27  109.47  306.27  158.67 1843.77  736.77
 1105.77]
4


Tylko 4 produkty kosztują powyżej 1000zł.

Zadanie 4: Filtrowanie

In [19]:
# Produkty z oceną >= 4.5 — nazwy i ceny
# Produkty poniżej 200 zł — nazwy i ilości w magazynie
# Podpowiedź: maska = oceny_klientow >= 4.5, potem ceny[maska]
# Uwaga: nazwy to zwykła lista — zamień ją na np.array(nazwy), żeby użyć maski

nazwy = np.array(nazwy)
maska = oceny_klientow >= 4.5
print(nazwy[maska])
print(ceny[maska])

maska = ceny < 200
print(nazwy[maska])
print(ilosci_magazyn[maska])

['Laptop' 'Smartfon' 'Słuchawki' 'Monitor']
[3499 2499  349 1499]
['Ładowarka' 'Mysz']
[312 201]


Produktami z oceną minimum 4.5 są laptop - 3499zł, smartfon - 2499zł, słuchawki - 349zł i monitor - 1499zł. 
Ładowarki i myszki kosztują poniżej 200zł. W magazynie znajduję sie 312 ładowarek i 201 myszek.

Zadanie 5: Analiza zapasów

In [27]:
# Produkty z niskim stanem magazynowym (< 30 szt.) — nazwy
# Łączna wartość tych produktów
# Ile % całego magazynu stanowią?
# Podpowiedź: wartosci = ceny * ilosci_magazyn, potem filtruj maską

wartosc_produktow = ceny * ilosci_magazyn

maska = ilosci_magazyn < 30
wartosc_stan = wartosc_produktow[maska].sum()

laczna_ilosc = ilosci_magazyn.sum()
ilosc = ilosci_magazyn[maska].sum()
procent = ilosc / laczna_ilosc * 100

print(nazwy[maska])
print(ilosci_magazyn[maska])
print(wartosc_stan)
print(procent)

['Laptop' 'Monitor']
[15 23]
86962
3.819095477386935


Laptopy i monitory są produktami z niskim stanem magazynowym, 15 i 23 sztuki. Ich łączna wartość wynosi 86962zł i stanowią około 4% całego magazynu.

Zadanie 6 (bonus): Analiza korelacji cena–ocena

In [29]:
# Które produkty są "okazją" — ocena >= 4.5 i cena < 1000 zł? Wypisz nazwy i ceny.
# Jaka jest średnia cena produktów z oceną >= 4.0 vs < 4.0?
# Podpowiedź: łączenie warunków: (warunek1) & (warunek2)

maska = (oceny_klientow >= 4.5) & (ceny < 1000)

okazje_nazwy = nazwy[maska]
okazje_ceny = ceny[maska]

print(okazje_ceny)
print(okazje_nazwy)

maska2 = oceny_klientow >= 4.0
maska3 = oceny_klientow < 4.0

srednia1 = ceny[maska2].mean()
srednia2 = ceny[maska3].mean()

print(srednia1)
print(srednia2)

[349]
['Słuchawki']
1446.142857142857
529.0


Okazją są słuchawki w cenie 349zł. Średnia cena produktów z oceną >= 4.00 wynosi około 1446zł, a z oceną < 4.0 - 529zł.

Ćwiczenie 4: Benchmark — lista vs NumPy

In [30]:
import time

In [31]:
# Przygotowanie danych
rozmiar = 1_000_000
lista = list(range(rozmiar))
tablica = np.array(lista)

In [32]:
# Test 1: Pomnóż przez 2
start = time.perf_counter()
wynik_lista = [x * 2 for x in lista]
czas_lista = time.perf_counter()- start
start = time.perf_counter()
wynik_numpy = tablica * 2
czas_numpy = time.perf_counter()- start
print(f"=== Mnożenie ×2 ({rozmiar:,} elementów) ===")
print(f"Lista: {czas_lista*1000:.1f} ms")
print(f"NumPy: {czas_numpy*1000:.1f} ms")
print(f"NumPy jest {czas_lista/czas_numpy:.0f}× szybszy!")

=== Mnożenie ×2 (1,000,000 elementów) ===
Lista: 46.1 ms
NumPy: 1.9 ms
NumPy jest 25× szybszy!


In [33]:
# Test 2: Suma
start = time.perf_counter()
suma_lista = sum(lista)
czas_lista = time.perf_counter()- start
start = time.perf_counter()
suma_numpy = tablica.sum()
czas_numpy = time.perf_counter()- start
print(f"\n=== Suma ({rozmiar:,} elementów) ===")
print(f"Lista: {czas_lista*1000:.1f} ms")
print(f"NumPy: {czas_numpy*1000:.1f} ms")
print(f"NumPy jest {czas_lista/czas_numpy:.0f}× szybszy!")


=== Suma (1,000,000 elementów) ===
Lista: 6.4 ms
NumPy: 2.3 ms
NumPy jest 3× szybszy!


In [34]:
# Test 3: Filtrowanie (elementy > 500000)
start = time.perf_counter()
filtr_lista = [x for x in lista if x > 500000]
czas_lista = time.perf_counter()- start
start = time.perf_counter()
filtr_numpy = tablica[tablica > 500000]
czas_numpy = time.perf_counter()- start
print(f"\n=== Filtrowanie > 500k ({rozmiar:,} elementów) ===")
print(f"Lista: {czas_lista*1000:.1f} ms")
print(f"NumPy: {czas_numpy*1000:.1f} ms")
print(f"NumPy jest {czas_lista/czas_numpy:.0f}× szybszy!")


=== Filtrowanie > 500k (1,000,000 elementów) ===
Lista: 27.7 ms
NumPy: 2.3 ms
NumPy jest 12× szybszy!


## Wnioski z benchmarku
- NumPy jest **25×** szybszy w mnożeniu
- NumPy jest **3×** szybszy w sumowaniu
- NumPy jest **12×** szybszy w filtrowaniu
- Wniosek: przy dużych danych **zawsze** używaj NumPy zamiast list